### PR TITLE
Ast refactor step 2

### PR DESCRIPTION
**Pull request by iamdefinitelyahuman**
### What I did
1. Refactor logic when generating AST nodes.
2. Add / expand docstrings using [NumPy style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html#example-numpy).

This is another step toward #1806.

### How I did it
1. `ast/utils.py`:
   a. Eliminate `parse_python_ast` and `_build_vyper_ast_init_kwargs` methods. Additional fields for the init kwargs is now added to the python nodes during annotation, and the logic around generating the vyper nodes happens within `VyperNode.__init__`.
2. `ast/nodes.py`:
   a. Individual nodes are now created using the `get_node` method. They can be generated from a python node, or dict.
   b. For compatibility with python3.8's  `Constant` node, I added a `_translated_fields` class attribute that remaps `value` to `n` or `s`.
   c. Nodes are converted to dicts using the `VyperNode.to_dict` method.

See the code for more detail.

### How to verify it
Run the tests.

### Cute Animal Picture
image

### CODE REVIEW

File: vyper/ast/annotation.py [hunk 1]
```diff
@@ -38,24 +62,42 @@ def visit_ClassDef(self, node):
...
+            raise SyntaxException(f"Invalid syntax (unsupported Python Constant AST node).", node)
+
+        return node
 
-class RewriteUnarySubVisitor(python_ast.NodeTransformer):
     def visit_UnaryOp(self, node):
         self.generic_visit(node)
-        if isinstance(node.op, python_ast.USub) and isinstance(node.operand, python_ast.Num):
+        # NOTE: This is done so that decimal literal now sees the negative sign as part of it
+        if isinstance(node.op, python_ast.USub) and isinstance(
```
Comment (charles-cooper) [inline vyper/ast/annotation.py]: stylistically I prefer to use variables to shorten the if condition:
```suggestion
is_usub = isinstance(node.op, python_ast.USub)
is_num = isinstance(node.operand, python_ast.Num)
if is_usub and is_num:
  ...
```

File: vyper/ast/annotation.py [hunk 2]
```diff
@@ -22,12 +32,26 @@ def __init__(self, source_code: str, class_types: Optional[ClassTypes] = None):
             self._class_types = {}
 
     def generic_visit(self, node):
-        # Decorate every node in the AST with the original source code. This is
-        # necessary to facilitate error pretty-printing.
+        # Decorate every node with the original source code to allow pretty-printing errors
         node.source_code = self._source_code
         node.node_id = self.counter
+        node.ast_type = node.__class__.__name__
```
Comment (fubuloubu) [inline vyper/ast/annotation.py]: Why not make this into a `class_name` property of `VyperNode`?

File: vyper/ast/nodes.py [hunk 3]
```diff
@@ -15,34 +14,107 @@
...
-    'src'
+    'lineno',
+    'node_id',
+    'source_code',
+    'src',
 )
+DICT_AST_SKIPLIST = ('source_code', )
+
+
+def get_node(ast_struct: typing.Union[typing.Dict, python_ast.AST]) -> "VyperNode":
```
Comment (fubuloubu) [inline vyper/ast/nodes.py]: I like this function

File: vyper/ast/nodes.py [hunk 4]
```diff
@@ -15,34 +14,107 @@
...
+            ast_struct
+        )
+
+    return vy_class(**ast_struct)
+
+
+def _to_node(value):
+    if isinstance(value, (dict, python_ast.AST)):
+        return get_node(value)
+    return value
```
Comment (fubuloubu) [inline vyper/ast/nodes.py]: Shouldn't this throw if conversion is not possible? Or do we not care because it's internal?

File: vyper/ast/nodes.py [hunk 5]
```diff
@@ -15,34 +14,107 @@
...
+
+
+def _to_dict(value):
+    if isinstance(value, VyperNode):
+        return value.to_dict()
+    return value
```
Comment (fubuloubu) [inline vyper/ast/nodes.py]: Same here as with `_to_node`

File: vyper/ast/nodes.py [hunk 6]
```diff
@@ -15,34 +14,107 @@
...
+    Vyper nodes are generated from, and closely resemble, their Python counterparts.
+
+    Attributes
+    ----------
+    __slots__ : Tuple
+        Allowed field names for the node.
+    _only_empty_fields : Tuple
+        Field names that, if present, must be set to None or a SyntaxException is
+        raised. This attribute is used to exclude syntax that is valid in Python
+        but not in vyper.
```
Comment (fubuloubu) [inline vyper/ast/nodes.py]: Why are these necessary? Because (for now) we're using Python AST nodes as a base class?

File: vyper/ast/nodes.py [hunk 7]
```diff
@@ -15,34 +14,107 @@
...
+    _translated_fields:
+        Field names that should be reassigned if encountered. Used to normalize
+        fields across different Python versions.
```
Comment (fubuloubu) [inline vyper/ast/nodes.py]: This goes away when we stop using Python AST base class?

File: vyper/ast/utils.py [hunk 8]
```diff
@@ -18,97 +18,25 @@
...
+    ----------
+    source_code : str
+        The vyper source code to parse.
+    source_id : int, optional
+        Source id to use in the .src member of each node.
+
+    Returns
+    -------
+    list
+        Generated vyper AST nodes.
```
Comment (fubuloubu) [inline vyper/ast/utils.py]: Typed or untyped? Optimized or un-optimized?

File: vyper/ast/utils.py [hunk 9]
```diff
@@ -117,70 +45,47 @@ def parse_to_ast(source_code: str, source_id: int = 0) -> list:
...
 
-
-@iterable_cast(dict)
-def _ast_to_dict(node: vy_ast.VyperNode) -> Generator:
-    for f in node.get_slots():
-        if f not in DICT_AST_SKIPLIST:
-            yield (f, ast_to_dict(getattr(node, f, None)))
-    yield ('ast_type', node.__class__.__name__)
+    # Convert to Vyper AST.
+    return vy_ast.get_node(py_ast).body  # type: ignore
```
Comment (fubuloubu) [inline vyper/ast/utils.py]: This recursively converts `py_ast` to `vy_ast`?

Also, eventually this should return the base AST node directly, right?

File: vyper/ast/nodes.py (hunk 4 above)
Comment (iamdefinitelyahuman) [reply vyper/ast/nodes.py]: This is me avoiding duplicated code, basically.  I pass _every_ member of the node through it, and if the value is a node, it gets converted.  Without the function I'd have to write some ugly code for evaluating items in lists.

File: vyper/ast/nodes.py (hunk 6 above)
Comment (iamdefinitelyahuman) [reply vyper/ast/nodes.py]: Yeah this is to block things like `else` within a for loop, or any other syntax that's valid in python but not in vyper. Once we do our own parsing it can go.

File: vyper/ast/nodes.py (hunk 7 above)
Comment (iamdefinitelyahuman) [reply vyper/ast/nodes.py]: Yes, this only exists to handle the differences between `Constant` and `Num`/`Str` with python 3.8. It was the cleanest approach I could come up with, under the hood `Constant` is providing `n` and `s` via a `@property` method so they don't show up in the `__dict__`.

File: vyper/ast/utils.py (hunk 9 above)
Comment (iamdefinitelyahuman) [reply vyper/ast/utils.py]: Yes, and yes. My next PR already has modified it to return `Module` instead of a list :)

Comment (iamdefinitelyahuman) [reply vyper/ast/utils.py]: ...but it would probably be better here anyway.

File: vyper/ast/annotation.py (hunk 2 above)
Comment (iamdefinitelyahuman) [reply vyper/ast/annotation.py]: It happens in the generic because it has to also happen in `visit_ClassDef` to handle python 3.8's `Constant` types.

File: vyper/ast/utils.py (hunk 8 above)
Comment (iamdefinitelyahuman) [reply vyper/ast/utils.py]: Expanded the comment.

File: vyper/ast/utils.py (hunk 9 above)
Comment (iamdefinitelyahuman) [reply vyper/ast/utils.py]: Done, and tests passing. I made `Module` iterable to reduce the pain of this change.
